rfcs/0013-ai-policies.md

```diff
@@ -0,0 +1,541 @@
+
+A Kuadrant `TokenRateLimitPolicy` is a custom resource provided by Kuadrant that targets Gateway API resources (`Gateway` and `HTTPRoute`) designed to allow users to define and enforce rate limiting rules to control token usage within OpenAI-style LLM provider APIs. It allows users to set and enforce token budget constraints for Gateways, but also for individual services exposed as HTTPRoutes. Per-user or per-group token rate limiting can be enforced based on JWT claims.
+
+
+### `PromptGuardPolicy`
+
+A Kuadrant `PromptGuardPolicy` is a custom resource provided by Kuadrant that targets Gateway API resources (`Gateway` and `HTTPRoute`), enabling users to define and enforce content safety rules with LLM prompts to detect and block sensitive prompts.  Prompt guards can be defined and enforced for both Gateways and individual HTTPRoutes.
+
+### `LLMResponseRiskCheckPolicy`
+
+A Kuadrant `LLMResponseRiskCheckPolicy` is a custom resource provided by Kuadrant that targets Gateway API resources (`Gateway` and `HTTPRoute`), enabling users to define and enforce content safety rules with LLM content completion responses, to detect and block sensitive responses from LLMs. These can be defined and enforced for both Gateways and individual HTTPRoutes.
+
+## Concepts Introduced
+
+- Token-based rate limiting: Controls usage by number of tokens, not just requests. Helps manage LLM costs.
+- Prompt guarding: Filters risky or unwanted prompts before they hit model servers.
+- Completion guarding: Reviews and filters model outputs before they're returned to users.
+
+## Example Use Case
+
+Say you're serving a chat LLM behind a Gateway, which implements an OpenAI-style chat `/completion` API. You want:
+
+- Free-tier users limited to 20k tokens/day.
+- Guardrails to reject prompts asking for harmful, violent or sexual content.
+- Output filtering to block generally harmful completions.
+
+You’d define the following policies and resources:
+
+> Notes (somewhere): OpenAI-style usage metrics in responses required: `"usage":{"prompt_tokens":5,"total_tokens":15,"completion_tokens":10}`
+
+```yaml
+apiVersion: kuadrant.io/v1
+kind: AuthPolicy
+metadata:
+  name: free-users-auth
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: Gateway
+    name: my-llm-gateway
+  rules:
+    authorization:
+      "free-users":
+        opa:
+          rego: |
+            groups := split(object.get(input.auth.identity.metadata.annotations, "kuadrant.io/groups", ""), ",")
+            allow { groups[_] == "free" }
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: api-key-free-user-1
+    app: my-llm
+  annotations:
+    kuadrant.io/groups: free
+    secret.kuadrant.io/user-id: user-1
+stringData:
+  api_key: iamafreeuser
+type: Opaque
+---
+apiVersion: kuadrant.io/v1alpha1
+kind: TokenRateLimitPolicy
+metadata:
+  name: token-limit-free
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: Gateway
+    name: my-llm-gateway
+  limit:
```

Comment:
an earlier version of this more closely mirrored `RateLimitPolicy`, in that multiple `limits` defined could be . Wondering why we diverged from that - simplicity? one for @david-martin later

cc @eguzki